pkg/virt-controller/watch/vm_initializer.go

```diff
@@ -0,0 +1,363 @@
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	kubev1 "kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/controller"
+	"kubevirt.io/kubevirt/pkg/kubecli"
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+type VirtualMachineInitializer struct {
+	vmPresetInformer cache.SharedIndexInformer
+	vmInitInformer   cache.SharedIndexInformer
+	clientset        kubecli.KubevirtClient
+	queue            workqueue.RateLimitingInterface
+	store            cache.Store
+}
+
+const initializerMarking = "presets.virtualmachines.kubevirt.io"
+
+func NewVirtualMachineInitializer(vmPresetInformer cache.SharedIndexInformer, vmInitInformer cache.SharedIndexInformer, queue workqueue.RateLimitingInterface, vmInitCache cache.Store, clientset kubecli.KubevirtClient) *VirtualMachineInitializer {
+	vmi := VirtualMachineInitializer{
+		vmPresetInformer: vmPresetInformer,
+		vmInitInformer:   vmInitInformer,
+		clientset:        clientset,
+		queue:            queue,
+		store:            vmInitCache,
+	}
+	return &vmi
+}
+
+func (c *VirtualMachineInitializer) Run(threadiness int, stopCh chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.queue.ShutDown()
+	log.Log.Info("Starting Virtual Machine Initializer.")
+
+	// Wait for cache sync before we start the pod controller
+	cache.WaitForCacheSync(stopCh, c.vmPresetInformer.HasSynced, c.vmInitInformer.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping controller.")
+}
+
+func (c *VirtualMachineInitializer) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *VirtualMachineInitializer) Execute() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	err := c.execute(key.(string))
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing VM %v", key)
+		c.queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed VM %v", key)
+		c.queue.Forget(key)
+	}
+	return true
+}
+
+func (c *VirtualMachineInitializer) execute(key string) error {
+
+	// Fetch the latest VM state from cache
+	obj, exists, err := c.store.GetByKey(key)
+
+	if err != nil {
+		return err
+	}
+
+	// If the VM isn't in the cache, it was just deleted, so shouldn't
+	// be initialized
+	if exists {
+		var vm *kubev1.VirtualMachine
+		vm = obj.(*kubev1.VirtualMachine)
+		// only process VM's that aren't initialized by this controller yet
+		if !isInitialized(vm) {
+			return c.initializeVirtualMachine(vm)
+		}
+	}
+
+	return nil
+}
+
+func (c *VirtualMachineInitializer) initializeVirtualMachine(vm *kubev1.VirtualMachine) error {
+	// All VM's must be marked as initialized or they are held in limbo forever
+	// Collect all errors and defer returning until after the update
+	logger := log.Log
+	errors := []error{}
```

Comment:
you don't need an error's slice here. The only place you're wanting to aggregate errors is during the applyPresets() function.

Reply:
example of the logic we should be replicating in the PodPreset code.
```
// safeToApplyPodPresetsOnPod determines if there is any conflict in information
// injected by given PodPresets in the Pod.
func safeToApplyPodPresetsOnPod(pod *api.Pod, podPresets []*settings.PodPreset) error {
        var errs []error

        // volumes attribute is defined at the Pod level, so determine if volumes
        // injection is causing any conflict.
        if _, err := mergeVolumes(pod.Spec.Volumes, podPresets); err != nil {
                errs = append(errs, err)
        }
        for _, ctr := range pod.Spec.Containers {
                if err := safeToApplyPodPresetsOnContainer(&ctr, podPresets); err != nil {
                        errs = append(errs, err)
                }
        }
        return utilerrors.NewAggregate(errs)
}
```

We should only be aggregating errors when we're looping through the presets

Reply:
Like you said elsewhere, we should only be returning errors (and thus rate limiting a re-try) when we end up with an error communicating with the Kubernetes API server. Thus the only time in applyPresets that we will have an error worth reporting is if we're unable to create an event to note a problem with presets.